### PR TITLE
ember-modal-dialog: Add types

### PR DESCRIPTION
**Pull request by tansongyang**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`. **Cannot set `strictFunctionTypes` because `ember` fails to compile with it.**

### CODE REVIEW

File: types/ember-modal-dialog/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6",
+            "dom"
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictNullChecks": true,
+        "strictFunctionTypes": false, // 'ember' types do not compile when true
```
Comment (mhegazy) [inline types/ember-modal-dialog/tsconfig.json]: please remove the comments, it is throwing the linter off.

Comment (tansongyang) [reply types/ember-modal-dialog/tsconfig.json]: Oops! Silly me. Thanks for the catch.

**Comment by tansongyang** [comment]
@mhegazy OK, so now it's failing because `ember-modal-dialog has unused file index.d.ts`. However, if I delete the file completely, the lint returns `no such file or directory` for `index.d.ts`. And then, if I add `index.d.ts` to the `tsconfig.json`, linting gives me `File has no content` on `index.d.ts`.

What do you recommend? It seems like this is a corner case in which there actually is nothing meaningful in `index.d.ts`.

**Comment by chriskrycho** [comment]
One alternative approach (which we've taken in the Ember.js proper typings) is, instead of using the files like this, to use `index.d.ts` and use `declare module` instead. That'd look roughly like this:

```ts
// Type definitions for ember-modal-dialog 2.4
// Project: https://github.com/yapplabs/ember-modal-dialog#readme
// Definitions by: Frank Tan <https://github.com/tansongyang>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 2.4

import Ember from 'ember';

declare module 'ember-modal-dialog/components/modal-dialog' {
  // all the type declarations currently in the types/ember-modal-dialog/components/modal-dialog.d.ts file
}
```

You should still be able to leave the test file as is, but it should resolve this issue.

**Comment by tansongyang** [comment]
@chriskrycho I believe I've tried this before, and I recall that the linter complained and said I should be using the file system instead of `declare module`. But I can try once more.

**Comment by tansongyang** [comment]
So the linter does indeed tell me to use the file system:

> ERROR: 9:16   no-declare-current-package       Instead of declaring a module with `declare module "ember-modal-dialog/components/modal-dialog"`, write its contents in directly in "ember-modal-dialog/components/modal-dialog.d.ts" or "ember-modal-dialog/components/modal-dialog/index.d.ts. See: https://github.com/Microsoft/dtslint/blob/master/docs/no-declare-current-package.md

**Comment by chriskrycho** [comment]
Yeah, you'll need to add an override `tslint.json` to support it one way or the other – either disabling the empty index lint, or disabling `no-declare-current-package`.

**Comment by chriskrycho** [comment]
For reference, see the tweaks we made for the Ember typings [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ember/tslint.json#L9-L13).

**Comment by tansongyang** [comment]
Ah, I was not aware that we were "allowed" to change lint rules. I'll have a fix later today.

On a side note, @mhegazy, do you have any idea why I didn't get the lint error about an empty `index.d.ts` in my local environment but got it in CI?

**Pull request merged**